src/components/encryption.ts

```diff
@@ -0,0 +1,275 @@
+            log.info("Got AS event before sync event");
+            // We weren't waiting for this event, but we might have got here too quick.
+            this.eventsPendingAS.push(event.event);
+            return;
+        }
+        const key = `${event.getRoomId()}:${event.getId()}`;
+        if (this.handledEvents.has(key)) {
+            // We're not interested in this event, as it's been handled.
+            return;
+        }
+        // First come, first serve handling.
+        this.handledEvents.add(key);
+        log.debug(`Handling ${event.getId()} through sync`);
+        this.onEvent(event.event);
+    }
+
+    private onTyping(syncUserId: string, event: any) {
+        if (this.userForRoom.get(event.getRoomId()) === syncUserId) {
+            // Ensure only the selected user for the room syncs this.
+            this.onEphemeralEvent(event.event);
+        }
+    }
+
+    private onReceipt(syncUserId: string, event: any) {
+        if (this.userForRoom.get(event.getRoomId()) === syncUserId) {
+            // Ensure only the user for the room syncs this.
+            this.onEphemeralEvent(event.event);
+        }
+    }
+
+    private onPresence(event: any) {
+        // Presence needs to be de-duplicated.
+        const now = Date.now();
+        const presenceEv = event.event as PresenceEvent;
+        const presenceContent = presenceEv.content;
+        const existingPresence = this.receivedPresence.find((p) =>
+            p.currentlyActive === presenceContent.currently_active &&
+            p.presence === presenceContent.presence &&
+            p.statusMsg === presenceContent.status_msg &&
+            p.userId === event.getSender()
+        );
+        if (existingPresence) {
+            // We've handled this already
+            return;
+        }
+        this.receivedPresence.push({
+            currentlyActive: presenceContent.currently_active,
+            presence: presenceContent.presence,
+            statusMsg: presenceContent.status_msg,
+            userId: event.getSender(),
+            ts: now,
+        });
+        this.onEphemeralEvent(presenceEv);
+    }
+
+    /**
+     * Start a sync loop for a given bridge user
+     * @param userId The user whos matrix client should start syncing
+     */
+    public async startSyncingUser(userId: string) {
+        log.info(`Starting to sync ${userId}`);
+        const intent = this.getIntent(userId);
+        await intent.ensureRegistered();
+        const matrixClient = intent.getClient();
+        matrixClient.on("event", this.onSyncEvent.bind(this));
+        matrixClient.on("error", (err: Error) => {
+            log.error(`${userId} client error:`, err);
+        });
+        matrixClient.on("RoomMember.typing", (event: TypingEvent) => this.onTyping(userId, event));
+        matrixClient.on("Room.receipt", (event: ReadReceiptEvent) => this.onReceipt(userId, event));
+        matrixClient.on("User.presence", this.onPresence.bind(this));
```

Comment:
Some methods need .bind() but onTyping and onReceipt don't?
Those methods use `this` to in their code.

Reply:
They use arrow functions which sidesteps the problem, but we should probably be consistent here.